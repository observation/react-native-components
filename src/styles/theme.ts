import Color from 'color'

const font = {
  size: {
    extraSmall: 10,
    small: 12,
    medium: 14,
    large: 16,
    extraLarge: 18,
    huge: 31,
  },
  lineHeight: {
    extraSmall: 16,
    small: 17,
    medium: 20,
    large: 26,
    extraLarge: 28,
    huge: 37,
  },
}

const icon = {
  size: {
    extraSmall: 10,
    small: 12,
    medium: 14,
    large: 16,
    extraLarge: 18,
    extraExtraLarge: 24,
    huge: 48,
  },
}

const black = '#212121'
const greyBase = '#939393'

const color = {
  primary: '#0066B1',
  primaryLighter: '#66A3D0',
  primaryLightest: '#E6F0F7',
  secondaryLight: '#709FFD',
  secondaryLightestExtra: '#F0F5FF',
  black,
  greyDark: '#666666',
  greyBase,
  placeholder: greyBase,
  greySemi: '#E6E6E6',
  greyLight: '#F8F9FA',
  white: '#FFFFFF',
  white70: '#FFFFFFB4',
  white77: '#FFFFFFC4',
  white80: '#FFFFFFCD',
  success: '#84B72C',
  successLight: '#9DC556',
  successLighter: '#B5D480',
  successLightest: '#CEE2AB',
  successLightestExtra: '#F7FBEF',
  successDark: '#6A9223',
  warning: '#F4C013',
  error: '#EA554B',
  errorLightest: '#F7BBB7',
  modalBackground: Color(black).alpha(0.6).lighten(0.4).string(),
  darkPanelBackground: Color(black).alpha(0.5).string(),
}

const margin = {
  eighth: 2,
  quarter: 4,
  half: 8,
  common: 16,
  large: 24,
  double: 32,
  huge: 48,
}

const rounded = {
  borderRadius: 4,
  overflow: 'hidden' as 'hidden',
}

const input = {
  ...rounded,
  flex: 1,
  minHeight: 40,
  borderWidth: 2,
  paddingLeft: margin.half,
  paddingRight: margin.double,
}

const bottomGradientColors = ['#30303000', '#30303059']
const topGradientColors = bottomGradientColors.slice().reverse()

const getBorderColor = ({ isFocused = false, hasErrors = false }) =>
  hasErrors ? color.error : isFocused ? color.primaryLighter : color.greySemi

export default {
  font,
  icon,
  color,
  bottomGradientColors,
  topGradientColors,
  margin,
  input,
  shadow: {
    android: {
      elevation: 6,
      shadowColor: color.greyBase,
    },
    ios: {
      shadowOffset: { width: 0, height: 4 },
      shadowColor: color.greySemi,
      shadowOpacity: 0.8,
      shadowRadius: 4,
    },
  },
  shadowSmall: {
    android: {
      elevation: 4,
      shadowColor: color.greyBase,
    },
    ios: {
      shadowOffset: { width: 0, height: 2 },
      shadowColor: color.greySemi,
      shadowOpacity: 0.8,
      shadowRadius: 2,
    },
  },
  rounded,
  roundedLarge: {
    borderRadius: 8,
    overflow: 'hidden' as 'hidden',
  },
  roundedHuge: {
    borderRadius: 10,
    overflow: 'hidden' as 'hidden',
  },
  absolute: {
    position: 'absolute' as 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  absoluteLeft: {
    position: 'absolute' as 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
  },
  absoluteRight: {
    position: 'absolute' as 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
  },
  absoluteBottom: {
    position: 'absolute' as 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
  },
  absoluteTop: {
    position: 'absolute' as 'absolute',
    top: 0,
    left: 0,
    right: 0,
  },
  getBorderColor,
}
