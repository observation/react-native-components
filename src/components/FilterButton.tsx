import React from 'react'
import { TouchableOpacity, StyleSheet, ViewStyle, StyleProp } from 'react-native'

import theme from '../styles/theme'

type Props = {
  content: (_: { color: string }) => JSX.Element
  onPress?: () => void
  active?: boolean
  style?: StyleProp<ViewStyle>
}

const FilterButton = ({ content, onPress, active = false, style }: Props) => {
  const buttonStyle = active ? styles.activeButton : styles.button
  const contentColor = active ? theme.color.white : theme.color.greyDark
  return (
    <TouchableOpacity
      testID="touchable-opacity"
      style={[styles.container, style, buttonStyle]}
      onPress={onPress}
      activeOpacity={0.5}
    >
      {content({ color: contentColor })}
    </TouchableOpacity>
  )
}

export default FilterButton

const styles = StyleSheet.create({
  container: {
    ...theme.rounded,
    justifyContent: 'center',
  },
  button: {
    backgroundColor: theme.color.white,
  },
  activeButton: {
    backgroundColor: theme.color.primary,
  },
})
