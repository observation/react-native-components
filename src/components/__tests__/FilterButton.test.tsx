import React from 'react'
import { Text } from 'react-native'

import { fireEvent, render } from '@testing-library/react-native'

import theme from '../../styles/theme'
import FilterButton from '../FilterButton'

const content = jest.fn(({}) => <Text>Some text</Text>)
const onPress = jest.fn()

describe('FilterButton', () => {
  describe('Rendering', () => {
    test('Enabled', () => {
      const { toJSON } = render(<FilterButton content={content} active onPress={onPress} style={{ flex: 1 }} />)
      expect(toJSON()).toMatchSnapshot()
      expect(content).toBeCalledWith({ color: theme.color.white })
    })

    test('Disabled', () => {
      const { toJSON } = render(<FilterButton content={content} active={false} style={{ flex: 1 }} />)
      expect(toJSON()).toMatchSnapshot()
      expect(content).toBeCalledWith({ color: theme.color.greyDark })
    })
  })

  describe('Interaction', () => {
    test('Click', async () => {
      const { getByText } = render(
        <FilterButton content={content} active={false} onPress={onPress} style={{ flex: 1 }} />,
      )
      await fireEvent.press(getByText('Some text'))
      expect(onPress).toBeCalled()
    })
  })
})
