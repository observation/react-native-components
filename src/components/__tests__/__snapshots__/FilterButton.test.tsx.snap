// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButton Rendering Disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 4,
      "flex": 1,
      "justifyContent": "center",
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <Text>
    Some text
  </Text>
</View>
`;

exports[`FilterButton Rendering Enabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#0066B1",
      "borderRadius": 4,
      "flex": 1,
      "justifyContent": "center",
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <Text>
    Some text
  </Text>
</View>
`;
